Rework Texture Interface

ebiederstadt opened the pull request:
Changes the texture interface use std::variant<> instead of individual load functions. This has the following benefits:
- Reduces code duplication. No need to specify GL_TEXTURE_2D (for example) in the constructor and the load function.
- Clearer user API. Users don't need to worry about which load function to call, they just pass in their data and that's it.
- Easy upgrade path. In the future we can add different texture types without adding another load function.

I have tested these changes in fork of bg-helloworld and they seem to be working nicely. If you like this direction I can have a PR ready for givderender quickly :)

ebiederstadt committed:
Rework Texture Interface

Changes the texture interface use std::variant<> instead of individual load functions. This has the following benfits:
- Reduces code duplication. No need to specify GL_TEXTURE_2D (for example) in the constructor and the load function.
- Clearer user API. Users don't need to worry about which load function to call, they just pass in their data and that's it.
- Easy upgrade path. In the future we can add different texture types without adding another load function.

lakinwecker reviewed (changes requested):
Marking this as "request changes" until we have time to discuss it (hopefully today).

My primary points are:
1. I don't think image loading should be part of this library. OpenGL doesn't care how you get your image, and I don't think poggles should either.
2. While this interface is unequivocally better than what we had before, it doesn't solve my primary issue with all of this, which is that we should not have to change this file to be able to change the way that poggles deals with textures. Ideally, poggles doesn't care _where_ the texture comes from. 
3. This all will be made moot by the changes I would like to propose that we make to givderender.

None of that prevents us from merging it, I'm happy to do so because of the fact that this improves the situation. I just want to make sure we do it with the above in mind.

JeremyAdamHart commented on source/poggles/texture.cpp:
```diff
@@ -1,94 +1,125 @@
+          },
+          [this](TextureFromData const& textureFromData)
+          {
+            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT);
+            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT);
+
+            GLenum format = GL_RGB;
+            if (textureFromData.channels == 1) {
+              format = GL_RED;
+            } else if (textureFromData.channels == 2) {
+              format = GL_RG;
+            } else if (textureFromData.channels == 3) {
+              format = GL_RGB;
+            } else if (textureFromData.channels == 4) {
+              format = GL_RGBA;
+            }
+            GLenum target = textureFromData.target.value_or(m_original_target);
```
m_originalTarget doesn't get updated in the texture loading, and there's no check to make sure that m_originalTarget is the same as GL_TEXTURE_2D. This is a problem in the original code, and definitely not something you introduced, but it's something we should probably fix :grimacing: 

```suggestion       
            m_originalTarget = textureFromData.target.value_or(m_original_target);

            glTexParameteri(m_originalTarget, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
            glTexParameteri(m_originalTarget, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
            glTexParameteri(m_originalTarget, GL_TEXTURE_WRAP_S, GL_REPEAT);
            glTexParameteri(m_originalTarget, GL_TEXTURE_WRAP_T, GL_REPEAT);

            GLenum format = GL_RGB;
            if (textureFromData.channels == 1) {
              format = GL_RED;
            } else if (textureFromData.channels == 2) {
              format = GL_RG;
            } else if (textureFromData.channels == 3) {
              format = GL_RGB;
            } else if (textureFromData.channels == 4) {
              format = GL_RGBA;
```

ebiederstadt replied:
This is out out date now :)

JeremyAdamHart reviewed (changes requested):
I think this is a fine way to do things, but I wonder, what are the advantages to doing this instead of having a function that creates (or updates) a texture?

There are many ways to load and create textures (For instance tex parameters can be set differently than how we currently have them set), and I'd have a preference of supporting those options outside of the details of the texture class. Instead of having these functions as a part of the texture class, we could have something like this:

`
loadTextureFromFile(poggles::texture& tex, std::filesystem::path filename){
    // Insert details here
}
`
Part of why I prefer this paradigm is that it can be done in any library, and the only thing that needs to change here is having the target made public. This adds some potential issues if people modify it randomly though which I don't love, but there may be a good way of handling that too

ebiederstadt commented:
There is no particular advantage with a function that creates a texture, I just went with this approach because it follows the class design that the rest of the library follows. I would be happy if we created functions to load textures instead of making users use the constructor (especially because it sounds like that is the direction that we will be taking this library anyways).

JeremyAdamHart committed:
Modification for texture PR

* fix: Fixing lints (#28)

Co-authored-by: Lakin Wecker <lakin@wecker.ca>

* v0.3.3 - fix: Fixing lints (#28)

Co-authored-by: Lakin Wecker <lakin@wecker.ca>

* Rework Texture Interface

Changes the texture interface use std::variant<> instead of individual load functions. This has the following benfits:
- Reduces code duplication. No need to specify GL_TEXTURE_2D (for example) in the constructor and the load function.
- Clearer user API. Users don't need to worry about which load function to call, they just pass in their data and that's it.
- Easy upgrade path. In the future we can add different texture types without adding another load function.

* Texture loading rework

* Use same syntax for input parameters that Lakin used

* Use target in glTexParameteri

* Move target assignment earlier

---------

Co-authored-by: Lakin Wecker <lakin.wecker@gmail.com>
Co-authored-by: Lakin Wecker <lakin@wecker.ca>
Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>
Co-authored-by: Erik Biederstadt <biederstadterik@gmail.com>

ebiederstadt commented:
Jeremy made all the changes we were talking about to this PR, so it should be good to go now :)

JeremyAdamHart reviewed (approved):
Yup! Let's merge it in!
Just some notes on the merge, this is a breaking change, so make sure you include feat!: at the beginning of the squashed commit message.
Additionally, since we updated the version.txt file, CI won't automatically do the version bump on its own, so you'll need to manually bump it to 0.3.4. Once you get that commit in it's all good to merge

benjaminulmer committed:
fix version

benjaminulmer committed:
Merge branch 'main' into improved_textures

benjaminulmer commented:
I went ahead and fixed the version so we can merge whenever now

lakinwecker commented on include/poggles/texture.h:
```diff
@@ -12,36 +15,30 @@ namespace poggles
-}  // namespace poggles
+
+auto loadFromFile(texture& tex,
+                  std::filesystem::path const& filepath,
```
I mildly prefer to use `poggles::texture::fromFile` as the name for these, but its a minor preference

ebiederstadt committed:
Updating PR feedback (#3)

* Rework Texture Interface

Changes the texture interface use std::variant<> instead of individual load functions. This has the following benfits:
- Reduces code duplication. No need to specify GL_TEXTURE_2D (for example) in the constructor and the load function.
- Clearer user API. Users don't need to worry about which load function to call, they just pass in their data and that's it.
- Easy upgrade path. In the future we can add different texture types without adding another load function.

* Texture loading rework

* Use same syntax for input parameters that Lakin used

* Use target in glTexParameteri

* Move target assignment earlier

* Lakin's feedback

---------

Co-authored-by: Jeremy Hart <jeremy@vividtheory.com>
